### PR TITLE
[RFR] Extract custom hooks from controllers

### PR DESCRIPTION
**Pull request by djhi**
Extract `useListParams` hook

### CODE REVIEW

File: packages/ra-core/src/controller/useList.ts [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        resource,
+        JSON.stringify(query),
+        JSON.stringify(location),
+    ];
+
+    const changeParams = useCallback(action => {
+        const newParams = queryReducer(query, action);
+        dispatch(
+            push({
+                ...location,
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/useList.ts]: I believe you don't have to pass the current location - the `history` package uses the current location by default. That way, if you add the `resource` to the function signature, you can pass an empty array as dependency arguments to `useCallback` and never worry about cache anymore.

File: packages/ra-core/src/controller/useList.ts [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+interface Actions {
+    changeParams: (action: any) => void;
+    setPage: (page: number) => void;
+    setPerPage: (pageSize: number) => void;
+    setSort: (sort: Sort) => void;
+}
+
+const useList = ({
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/useList.ts]: I'd call it useListParams

**Comment by fzaninotto** [comment]
Why didn't you move setFilter()?

**Comment by djhi** [comment]
> Why didn't you move setFilter()?

I was planning to extract a separate hook for filters but it may be too soon. I'll try to transform the controller into a single hook first

File: packages/ra-core/src/controller/useList.ts (hunk 2 above)
Comment (djhi) [reply packages/ra-core/src/controller/useList.ts]: `useCrudList` ?

Comment (fzaninotto) [reply packages/ra-core/src/controller/useList.ts]: we tend to use the Crud word for communication with the dataProvider. this hook does not communicate with the dataProvider, so I'm -1 for that name.

File: packages/ra-core/src/controller/useListParams.ts [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    filterDefaultValues,
+    sort = {
+        field: 'id',
+        order: SORT_ASC,
+    },
+    perPage = 10,
+    filter,
+    debounce = 500,
+}: Props): [Query, Actions] => {
+    const [displayedFilters, setDisplayedFilters] = useState({});
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/useListParams.ts]: This hoos should only contain the query-related logic, so I think that displayed filters, data from the provider, and the initial effect should stay in the controller (or go to another hook).

File: packages/ra-core/src/controller/useListParams.ts [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    changeParams: (action: any) => void;
+    setPage: (page: number) => void;
+    setPerPage: (pageSize: number) => void;
+    setSort: (sort: Sort) => void;
+    setFilters: (filters: any) => void;
+    hideFilter: (filterName: string) => void;
+    showFilter: (filterName: string, defaultValue: any) => void;
+}
+
+const useListParams = ({
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/useListParams.ts]: I'd love a small jsDoc explaining the return value

File: packages/ra-core/src/controller/useListParams.ts (hunk 3 above)
Comment (djhi) [reply packages/ra-core/src/controller/useListParams.ts]: About the filters, that would mean moving the filters actions out too

Comment (fzaninotto) [reply packages/ra-core/src/controller/useListParams.ts]: right

Comment (djhi) [reply packages/ra-core/src/controller/useListParams.ts]: Maybe too soon to do that ? What about renaming them `activeFilters` and `setActiveFilters` ?

File: packages/ra-core/src/controller/useListParams.ts (hunk 4 above)
Comment (djhi) [reply packages/ra-core/src/controller/useListParams.ts]: Added some

**Pull request merged**